File: mobile/ios/NavTin/Engine/AIModule.swift
import Foundation

class AIModule {
    struct MoveAction {
        let fromType: GamePhase
        let nodeIndex: Int
        let fromIndex: Int?
    }

    static func getNextMove(state: GameState) -> (Int, Int?)? {
        let difficulty = state.aiDifficulty
        let depth: Int
        
        switch difficulty {
        case .easy: depth = 1
        case .medium: depth = 3
        case .hard: depth = 5
        }
        
        let bestMove = minimax(state: state, depth: depth, isMaximizing: true, alpha: -1000000, beta: 1000000)
        return bestMove.action
    }
    
    // Simplified heuristic function
    private static func evaluate(state: GameState) -> Int {
        let aiPieces = state.player2PiecesOnBoard
        let humanPieces = state.player1PiecesOnBoard
        
        var score = (aiPieces - humanPieces) * 100
        
        // Bonus for mills
        let aiMills = countMills(for: .player2, positions: state.piecePositions)
        let humanMills = countMills(for: .player1, positions: state.piecePositions)
        score += (aiMills - humanMills) * 50
        
        // Mobility bonus
        let aiMoves = countPossibleMoves(for: .player2, state: state)
        let humanMoves = countPossibleMoves(for: .player1, state: state)
        score += (aiMoves - humanMoves) * 10
        
        return score
    }
    
    private static func minimax(state: GameState, depth: Int, isMaximizing: Bool, alpha: Int, beta: Int) -> (score: Int, action: (Int, Int?)?) {
        if depth == 0 || isGameOver(state) {
            return (evaluate(state: state), nil)
        }
        
        var currentAlpha = alpha
        var currentBeta = beta
        
        if isMaximizing {
            var maxEval = -1000000
            var bestAction: (Int, Int?)? = nil
            
            let possibleMoves = getAllMoves(for: .player2, state: state)
            for move in possibleMoves {
                let nextState = simulateMove(move, state: state)
                let eval = minimax(state: nextState, depth: depth - 1, isMaximizing: false, alpha: currentAlpha, beta: currentBeta).score
                if eval > maxEval {
                    maxEval = eval
                    bestAction = move
                }
                currentAlpha = max(currentAlpha, eval)
                if currentBeta <= currentAlpha { break }
            }
            return (maxEval, bestAction)
        } else {
            var minEval = 1000000
            var bestAction: (Int, Int?)? = nil
            
            let possibleMoves = getAllMoves(for: .player1, state: state)
            for move in possibleMoves {
                let nextState = simulateMove(move, state: state)
                let eval = minimax(state: nextState, depth: depth - 1, isMaximizing: true, alpha: currentAlpha, beta: currentBeta).score
                if eval < minEval {
                    minEval = eval
                    bestAction = move
                }
                currentBeta = min(currentBeta, eval)
                if currentBeta <= currentAlpha { break }
            }
            return (minEval, bestAction)
        }
    }
    
    // Helpers (Internal simulation logic)
    private static func getAllMoves(for player: PlayerTurn, state: GameState) -> [(Int, Int?)] {
        var moves: [(Int, Int?)] = []
        let type = player == .player1 ? PieceType.player1 : PieceType.player2
        
        if state.phase == .placing {
            for i in 0...23 {
                if state.piecePositions[i] == nil { moves.append((i, nil)) }
            }
        } else {
            let playerIndices = state.piecePositions.filter { $0.value == type }.map { $0.key }
            for fromIdx in playerIndices {
                if state.phase == .flying || (player == .player1 ? state.player1PiecesOnBoard == 3 : state.player2PiecesOnBoard == 3) {
                    for toIdx in 0...23 {
                        if state.piecePositions[toIdx] == nil { moves.append((toIdx, fromIdx)) }
                    }
                } else {
                    for toIdx in Board.adjacencyList[fromIdx] ?? [] {
                        if state.piecePositions[toIdx] == nil { moves.append((toIdx, fromIdx)) }
                    }
                }
            }
        }
        return moves
    }
    
    private static func simulateMove(_ move: (Int, Int?), state: GameState) -> GameState {
        // Deep copy of state for simulation
        let newState = GameState()
        newState.piecePositions = state.piecePositions
        newState.player1PiecesOnBoard = state.player1PiecesOnBoard
        newState.player2PiecesOnBoard = state.player2PiecesOnBoard
        newState.player1PiecesLeftToPlace = state.player1PiecesLeftToPlace
        newState.player2PiecesLeftToPlace = state.player2PiecesLeftToPlace
        newState.phase = state.phase
        newState.turn = state.turn
        
        let nodeIndex = move.0
        let fromIndex = move.1
        let type = newState.turn == .player1 ? PieceType.player1 : PieceType.player2
        
        if let from = fromIndex {
            newState.piecePositions[from] = nil
        } else {
            if newState.turn == .player1 { newState.player1PiecesLeftToPlace -= 1; newState.player1PiecesOnBoard += 1 }
            else { newState.player2PiecesLeftToPlace -= 1; newState.player2PiecesOnBoard += 1 }
        }
        newState.piecePositions[nodeIndex] = type
        
        // Handle mill removal in simulation (very simple: remove closest opponent piece)
        if MillDetector.isPartOfMill(nodeIndex: nodeIndex, positions: newState.piecePositions, player: type) {
            let opponent = newState.turn == .player1 ? PieceType.player2 : PieceType.player1
            if let targetIdx = newState.piecePositions.first(where: { $0.value == opponent })?.key {
                newState.piecePositions[targetIdx] = nil
                if opponent == .player1 { newState.player1PiecesOnBoard -= 1 }
                else { newState.player2PiecesOnBoard -= 1 }
            }
        }
        
        newState.turn = newState.turn == .player1 ? .player2 : .player1
        if newState.player1PiecesLeftToPlace == 0 && newState.player2PiecesLeftToPlace == 0 {
            newState.phase = .moving
        }
        return newState
    }
    
    private static func countMills(for player: PlayerTurn, positions: [Int: PieceType]) -> Int {
        let type = player == .player1 ? PieceType.player1 : PieceType.player2
        var count = 0
        for mill in StandardMills.all {
            if mill.allSatisfy({ positions[$0] == type }) { count += 1 }
        }
        return count
    }
    
    private static func countPossibleMoves(for player: PlayerTurn, state: GameState) -> Int {
        return getAllMoves(for: player, state: state).count
    }
    
    private static func isGameOver(_ state: GameState) -> Bool {
        if state.player1PiecesLeftToPlace > 0 || state.player2PiecesLeftToPlace > 0 { return false }
        return state.player1PiecesOnBoard < 3 || state.player2PiecesOnBoard < 3
    }
}

File: mobile/android/app/src/main/java/com/navtin/engine/AIModule.kt
package com.navtin.engine

import kotlin.math.max
import kotlin.math.min

object AIModule {
    fun getNextMove(state: GameState): Pair<Int, Int?>? {
        val depth = when (state.aiDifficulty) {
            AIDifficulty.EASY -> 1
            AIDifficulty.MEDIUM -> 3
            AIDifficulty.HARD -> 5
        }
        
        val result = minimax(state, depth, true, -1000000, 1000000)
        return result.second
    }

    private fun evaluate(state: GameState): Int {
        val aiPieces = state.player2PiecesOnBoard
        val humanPieces = state.player1PiecesOnBoard
        var score = (aiPieces - humanPieces) * 100
        
        val aiMills = countMills(PlayerTurn.PLAYER_2, state.piecePositions)
        val humanMills = countMills(PlayerTurn.PLAYER_1, state.piecePositions)
        score += (aiMills - humanMills) * 50
        
        val aiMoves = countPossibleMoves(PlayerTurn.PLAYER_2, state)
        val humanMoves = countPossibleMoves(PlayerTurn.PLAYER_1, state)
        score += (aiMoves - humanMoves) * 10
        
        return score
    }

    private fun minimax(state: GameState, depth: Int, isMaximizing: Boolean, alpha: Int, beta: Int): Pair<Int, Pair<Int, Int?>?> {
        if (depth == 0 || isGameOver(state)) {
            return evaluate(state) to null
        }
        
        var currentAlpha = alpha
        var currentBeta = beta
        
        if (isMaximizing) {
            var maxEval = -1000000
            var bestAction: Pair<Int, Int?>? = null
            
            for (move in getAllMoves(PlayerTurn.PLAYER_2, state)) {
                val nextState = simulateMove(move, state)
                val eval = minimax(nextState, depth - 1, false, currentAlpha, currentBeta).first
                if (eval > maxEval) {
                    maxEval = eval
                    bestAction = move
                }
                currentAlpha = max(currentAlpha, eval)
                if (currentBeta <= currentAlpha) break
            }
            return maxEval to bestAction
        } else {
            var minEval = 1000000
            var bestAction: Pair<Int, Int?>? = null
            
            for (move in getAllMoves(PlayerTurn.PLAYER_1, state)) {
                val nextState = simulateMove(move, state)
                val eval = minimax(nextState, depth - 1, true, currentAlpha, currentBeta).first
                if (eval < minEval) {
                    minEval = eval
                    bestAction = move
                }
                currentBeta = min(currentBeta, eval)
                if (currentBeta <= currentAlpha) break
            }
            return minEval to bestAction
        }
    }

    private fun getAllMoves(player: PlayerTurn, state: GameState): List<Pair<Int, Int?>> {
        val moves = mutableListOf<Pair<Int, Int?>>()
        val type = if (player == PlayerTurn.PLAYER_1) PieceType.PLAYER_1 else PieceType.PLAYER_2
        
        if (state.phase == GamePhase.PLACING) {
            for (i in 0..23) {
                if (!state.piecePositions.containsKey(i)) moves.add(i to null)
            }
        } else {
            val playerIndices = state.piecePositions.filter { it.value == type }.keys
            for (fromIdx in playerIndices) {
                val canFly = (player == PlayerTurn.PLAYER_1 && state.player1PiecesOnBoard == 3) ||
                             (player == PlayerTurn.PLAYER_2 && state.player2PiecesOnBoard == 3)
                
                if (canFly) {
                    for (toIdx in 0..23) {
                        if (!state.piecePositions.containsKey(toIdx)) moves.add(toIdx to fromIdx)
                    }
                } else {
                    Board.adjacencyList[fromIdx]?.forEach { toIdx ->
                        if (!state.piecePositions.containsKey(toIdx)) moves.add(toIdx to fromIdx)
                    }
                }
            }
        }
        return moves
    }

    private fun simulateMove(move: Pair<Int, Int?>, state: GameState): GameState {
        val newState = GameState()
        // Manual deep copy of mapping
        state.piecePositions.forEach { (k, v) -> newState.piecePositions[k] = v }
        newState.player1PiecesOnBoard = state.player1PiecesOnBoard
        newState.player2PiecesOnBoard = state.player2PiecesOnBoard
        newState.player1PiecesLeftToPlace = state.player1PiecesLeftToPlace
        newState.player2PiecesLeftToPlace = state.player2PiecesLeftToPlace
        newState.phase = state.phase
        newState.turn = state.turn
        
        val nodeIndex = move.first
        val fromIndex = move.second
        val type = if (newState.turn == PlayerTurn.PLAYER_1) PieceType.PLAYER_1 else PieceType.PLAYER_2
        
        if (fromIndex != null) {
            newState.piecePositions.remove(fromIndex)
        } else {
            if (newState.turn == PlayerTurn.PLAYER_1) {
                newState.player1PiecesLeftToPlace--
                newState.player1PiecesOnBoard++
            } else {
                newState.player2PiecesLeftToPlace--
                newState.player2PiecesOnBoard++
            }
        }
        newState.piecePositions[nodeIndex] = type
        
        if (MillDetector.isPartOfMill(nodeIndex, newState.piecePositions, type)) {
            val opponent = if (newState.turn == PlayerTurn.PLAYER_1) PieceType.PLAYER_2 else PieceType.PLAYER_1
            val targetIdx = newState.piecePositions.entries.firstOrNull { it.value == opponent }?.key
            if (targetIdx != null) {
                newState.piecePositions.remove(targetIdx)
                if (opponent == PieceType.PLAYER_1) newState.player1PiecesOnBoard--
                else newState.player2PiecesOnBoard--
            }
        }
        
        newState.turn = if (newState.turn == PlayerTurn.PLAYER_1) PlayerTurn.PLAYER_2 else PlayerTurn.PLAYER_1
        if (newState.player1PiecesLeftToPlace == 0 && newState.player2PiecesLeftToPlace == 0) {
            newState.phase = GamePhase.MOVING
        }
        return newState
    }

    private fun countMills(player: PlayerTurn, positions: Map<Int, PieceType>): Int {
        val type = if (player == PlayerTurn.PLAYER_1) PieceType.PLAYER_1 else PieceType.PLAYER_2
        return StandardMills.all.count { mill -> mill.all { positions[it] == type } }
    }

    private fun countPossibleMoves(player: PlayerTurn, state: GameState): Int {
        return getAllMoves(player, state).size
    }

    private fun isGameOver(state: GameState): Boolean {
        if (state.player1PiecesLeftToPlace > 0 || state.player2PiecesLeftToPlace > 0) return false
        if (state.player1PiecesOnBoard < 3 || state.player2PiecesOnBoard < 3) return true
        if (countPossibleMoves(state.turn, state) == 0) return true
        return false
    }
}
